File: packages/java-sdk/src/main/java/com/turbodocx/TurboPartner.java
package com.turbodocx;

import com.google.gson.JsonObject;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * TurboPartner module for partner portal management operations.
 *
 * <p>Provides partner-level management for multi-tenant applications:</p>
 * <ul>
 *   <li>Organization management (CRUD, entitlements)</li>
 *   <li>Organization user management</li>
 *   <li>Organization API key management</li>
 *   <li>Partner API key management</li>
 *   <li>Partner portal user management</li>
 *   <li>Audit logs</li>
 * </ul>
 *
 * <p>All methods return {@link JsonObject} containing the raw API response
 * with {@code success}, {@code data}, and optionally {@code message} fields.</p>
 */
public class TurboPartner {

    private final PartnerHttpClient httpClient;
    private final String partnerId;

    TurboPartner(PartnerHttpClient httpClient, String partnerId) {
        this.httpClient = httpClient;
        this.partnerId = partnerId;
    }

    private String basePath() {
        return "/partner/" + partnerId;
    }

    // =========================================================================
    // Organization Management
    // =========================================================================

    /**
     * Create a new organization under the partner account.
     *
     * @param name     Organization name (required)
     * @param metadata Optional metadata key-value pairs (may be null)
     * @param features Optional entitlement configuration (may be null)
     * @return JsonObject with success and data (organization details)
     * @throws IOException if the request fails
     */
    public JsonObject createOrganization(String name, Map<String, Object> metadata, Map<String, Object> features) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        if (metadata != null) {
            body.put("metadata", metadata);
        }
        if (features != null) {
            body.put("features", features);
        }
        return httpClient.post(basePath() + "/organization", body);
    }

    /**
     * Create a new organization with just a name.
     *
     * @param name Organization name (required)
     * @return JsonObject with success and data (organization details)
     * @throws IOException if the request fails
     */
    public JsonObject createOrganization(String name) throws IOException {
        return createOrganization(name, null, null);
    }

    /**
     * List all organizations under the partner account.
     *
     * @param limit  Maximum number of results (may be null)
     * @param offset Pagination offset (may be null)
     * @param search Search query string (may be null)
     * @return JsonObject with success and data (results, totalRecords, limit, offset)
     * @throws IOException if the request fails
     */
    public JsonObject listOrganizations(Integer limit, Integer offset, String search) throws IOException {
        String qs = buildQueryString(
                "limit", limit,
                "offset", offset,
                "search", search
        );
        return httpClient.get(basePath() + "/organizations" + qs);
    }

    /**
     * Get full organization details including features and tracking.
     *
     * @param organizationId Organization UUID
     * @return JsonObject with success and data (organization info, features, tracking)
     * @throws IOException if the request fails
     */
    public JsonObject getOrganizationDetails(String organizationId) throws IOException {
        return httpClient.get(basePath() + "/organizations/" + organizationId);
    }

    /**
     * Update an organization's name.
     *
     * @param organizationId Organization UUID
     * @param name           New organization name
     * @return JsonObject with success and data (updated organization)
     * @throws IOException if the request fails
     */
    public JsonObject updateOrganizationInfo(String organizationId, String name) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        return httpClient.patch(basePath() + "/organizations/" + organizationId, body);
    }

    /**
     * Delete an organization.
     *
     * @param organizationId Organization UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject deleteOrganization(String organizationId) throws IOException {
        return httpClient.delete(basePath() + "/organizations/" + organizationId);
    }

    /**
     * Update organization entitlements (feature limits and capabilities).
     *
     * @param organizationId Organization UUID
     * @param features       Feature configuration (may be null)
     * @param tracking       Tracking data (may be null)
     * @return JsonObject with success and data (features, tracking)
     * @throws IOException if the request fails
     */
    public JsonObject updateOrganizationEntitlements(String organizationId, Map<String, Object> features, Map<String, Object> tracking) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        if (features != null) {
            body.put("features", features);
        }
        if (tracking != null) {
            body.put("tracking", tracking);
        }
        return httpClient.patch(basePath() + "/organizations/" + organizationId + "/entitlements", body);
    }

    // =========================================================================
    // Organization User Management
    // =========================================================================

    /**
     * List users in an organization.
     *
     * @param organizationId Organization UUID
     * @param limit          Maximum number of results (may be null)
     * @param offset         Pagination offset (may be null)
     * @param search         Search query string (may be null)
     * @return JsonObject with success and data (results, totalRecords, limit, offset)
     * @throws IOException if the request fails
     */
    public JsonObject listOrganizationUsers(String organizationId, Integer limit, Integer offset, String search) throws IOException {
        String qs = buildQueryString(
                "limit", limit,
                "offset", offset,
                "search", search
        );
        return httpClient.get(basePath() + "/organizations/" + organizationId + "/users" + qs);
    }

    /**
     * Add a user to an organization.
     *
     * @param organizationId Organization UUID
     * @param email          User email address
     * @param role           User role ("admin", "contributor", "user", or "viewer")
     * @return JsonObject with success and data (user details)
     * @throws IOException if the request fails
     */
    public JsonObject addUserToOrganization(String organizationId, String email, String role) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("email", email);
        body.put("role", role);
        return httpClient.post(basePath() + "/organizations/" + organizationId + "/users", body);
    }

    /**
     * Update a user's role in an organization.
     *
     * @param organizationId Organization UUID
     * @param userId         User UUID
     * @param role           New role ("admin", "contributor", "user", or "viewer")
     * @return JsonObject with success and data (updated user)
     * @throws IOException if the request fails
     */
    public JsonObject updateOrganizationUserRole(String organizationId, String userId, String role) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("role", role);
        return httpClient.patch(basePath() + "/organizations/" + organizationId + "/users/" + userId, body);
    }

    /**
     * Remove a user from an organization.
     *
     * @param organizationId Organization UUID
     * @param userId         User UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject removeUserFromOrganization(String organizationId, String userId) throws IOException {
        return httpClient.delete(basePath() + "/organizations/" + organizationId + "/users/" + userId);
    }

    /**
     * Resend invitation email to a user in an organization.
     *
     * @param organizationId Organization UUID
     * @param userId         User UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject resendOrganizationInvitationToUser(String organizationId, String userId) throws IOException {
        return httpClient.post(basePath() + "/organizations/" + organizationId + "/users/" + userId + "/resend-invitation", null);
    }

    // =========================================================================
    // Organization API Key Management
    // =========================================================================

    /**
     * List API keys for an organization.
     *
     * @param organizationId Organization UUID
     * @param limit          Maximum number of results (may be null)
     * @param offset         Pagination offset (may be null)
     * @param search         Search query string (may be null)
     * @return JsonObject with success and data (results, totalRecords, limit, offset)
     * @throws IOException if the request fails
     */
    public JsonObject listOrganizationApiKeys(String organizationId, Integer limit, Integer offset, String search) throws IOException {
        String qs = buildQueryString(
                "limit", limit,
                "offset", offset,
                "search", search
        );
        return httpClient.get(basePath() + "/organizations/" + organizationId + "/apikeys" + qs);
    }

    /**
     * Create an API key for an organization.
     *
     * @param organizationId Organization UUID
     * @param name           API key name
     * @param role           API key role ("admin", "contributor", or "viewer")
     * @return JsonObject with success, data (key details including the key value), and message
     * @throws IOException if the request fails
     */
    public JsonObject createOrganizationApiKey(String organizationId, String name, String role) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        body.put("role", role);
        return httpClient.post(basePath() + "/organizations/" + organizationId + "/apikeys", body);
    }

    /**
     * Update an organization API key.
     *
     * @param organizationId Organization UUID
     * @param apiKeyId       API key UUID
     * @param name           New API key name (may be null to keep current)
     * @param role           New API key role (may be null to keep current)
     * @return JsonObject with success, message, and apiKey (updated key details)
     * @throws IOException if the request fails
     */
    public JsonObject updateOrganizationApiKey(String organizationId, String apiKeyId, String name, String role) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        if (name != null) {
            body.put("name", name);
        }
        if (role != null) {
            body.put("role", role);
        }
        return httpClient.patch(basePath() + "/organizations/" + organizationId + "/apikeys/" + apiKeyId, body);
    }

    /**
     * Revoke (delete) an organization API key.
     *
     * @param organizationId Organization UUID
     * @param apiKeyId       API key UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject revokeOrganizationApiKey(String organizationId, String apiKeyId) throws IOException {
        return httpClient.delete(basePath() + "/organizations/" + organizationId + "/apikeys/" + apiKeyId);
    }

    // =========================================================================
    // Partner API Key Management
    // =========================================================================

    /**
     * List partner-level API keys.
     *
     * @param limit  Maximum number of results (may be null)
     * @param offset Pagination offset (may be null)
     * @param search Search query string (may be null)
     * @return JsonObject with success and data (results, totalRecords, limit, offset)
     * @throws IOException if the request fails
     */
    public JsonObject listPartnerApiKeys(Integer limit, Integer offset, String search) throws IOException {
        String qs = buildQueryString(
                "limit", limit,
                "offset", offset,
                "search", search
        );
        return httpClient.get(basePath() + "/api-keys" + qs);
    }

    /**
     * Create a partner-level API key with scoped permissions.
     *
     * @param name        API key name
     * @param scopes      List of permission scopes (use {@link PartnerScope} constants)
     * @param description Optional description (may be null)
     * @return JsonObject with success, data (key details including the key value), and message
     * @throws IOException if the request fails
     */
    public JsonObject createPartnerApiKey(String name, List<String> scopes, String description) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        body.put("scopes", scopes);
        if (description != null) {
            body.put("description", description);
        }
        return httpClient.post(basePath() + "/api-keys", body);
    }

    /**
     * Update a partner API key.
     *
     * @param keyId       API key UUID
     * @param name        New name (may be null to keep current)
     * @param description New description (may be null to keep current)
     * @param scopes      New scopes list (may be null to keep current)
     * @return JsonObject with success, message, and apiKey (updated key details)
     * @throws IOException if the request fails
     */
    public JsonObject updatePartnerApiKey(String keyId, String name, String description, List<String> scopes) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        if (name != null) {
            body.put("name", name);
        }
        if (description != null) {
            body.put("description", description);
        }
        if (scopes != null) {
            body.put("scopes", scopes);
        }
        return httpClient.patch(basePath() + "/api-keys/" + keyId, body);
    }

    /**
     * Revoke (delete) a partner API key.
     *
     * @param keyId API key UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject revokePartnerApiKey(String keyId) throws IOException {
        return httpClient.delete(basePath() + "/api-keys/" + keyId);
    }

    // =========================================================================
    // Partner Portal User Management
    // =========================================================================

    /**
     * List users in the partner portal.
     *
     * @param limit  Maximum number of results (may be null)
     * @param offset Pagination offset (may be null)
     * @param search Search query string (may be null)
     * @return JsonObject with success and data (results, totalRecords, limit, offset)
     * @throws IOException if the request fails
     */
    public JsonObject listPartnerPortalUsers(Integer limit, Integer offset, String search) throws IOException {
        String qs = buildQueryString(
                "limit", limit,
                "offset", offset,
                "search", search
        );
        return httpClient.get(basePath() + "/users" + qs);
    }

    /**
     * Add a user to the partner portal.
     *
     * @param email       User email address
     * @param role        User role ("admin", "member", or "viewer")
     * @param permissions Permission flags with camelCase keys
     * @return JsonObject with success and data (user details)
     * @throws IOException if the request fails
     */
    public JsonObject addUserToPartnerPortal(String email, String role, Map<String, Boolean> permissions) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("email", email);
        body.put("role", role);
        body.put("permissions", permissions);
        return httpClient.post(basePath() + "/users", body);
    }

    /**
     * Update a partner portal user's role and/or permissions.
     *
     * @param userId      User UUID
     * @param role        New role (may be null to keep current)
     * @param permissions New permission flags (may be null to keep current)
     * @return JsonObject with success and data (userId, role, permissions)
     * @throws IOException if the request fails
     */
    public JsonObject updatePartnerUserPermissions(String userId, String role, Map<String, Boolean> permissions) throws IOException {
        Map<String, Object> body = new LinkedHashMap<>();
        if (role != null) {
            body.put("role", role);
        }
        if (permissions != null) {
            body.put("permissions", permissions);
        }
        return httpClient.patch(basePath() + "/users/" + userId, body);
    }

    /**
     * Remove a user from the partner portal.
     *
     * @param userId User UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject removeUserFromPartnerPortal(String userId) throws IOException {
        return httpClient.delete(basePath() + "/users/" + userId);
    }

    /**
     * Resend partner portal invitation email to a user.
     *
     * @param userId User UUID
     * @return JsonObject with success and message
     * @throws IOException if the request fails
     */
    public JsonObject resendPartnerPortalInvitationToUser(String userId) throws IOException {
        return httpClient.post(basePath() + "/users/" + userId + "/resend-invitation", null);
    }

    // =========================================================================
    // Audit Logs
    // =========================================================================

    /**
     * Get partner audit logs with optional filtering.
     *
     * @param limit        Maximum number of results (may be null)
     * @param offset       Pagination offset (may be null)
     * @param search       Search query string (may be null)
     * @param action       Filter by action type (may be null)
     * @param resourceType Filter by resource type (may be null)
     * @param resourceId   Filter by resource ID (may be null)
     * @param success      Filter by success status (may be null)
     * @param startDate    Filter from date, ISO 8601 (may be null)
     * @param endDate      Filter to date, ISO 8601 (may be null)
     * @return JsonObject with success and data (results, totalRecords, limit, offset)
     * @throws IOException if the request fails
     */
    public JsonObject getPartnerAuditLogs(
            Integer limit, Integer offset, String search,
            String action, String resourceType, String resourceId,
            Boolean success, String startDate, String endDate
    ) throws IOException {
        String qs = buildQueryString(
                "limit", limit,
                "offset", offset,
                "search", search,
                "action", action,
                "resourceType", resourceType,
                "resourceId", resourceId,
                "success", success,
                "startDate", startDate,
                "endDate", endDate
        );
        return httpClient.get(basePath() + "/audit-logs" + qs);
    }

    // =========================================================================
    // Helpers
    // =========================================================================

    /**
     * Build URL query string from key-value pairs, skipping null values.
     * Accepts alternating key (String) and value (Object) arguments.
     */
    private static String buildQueryString(Object... params) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < params.length - 1; i += 2) {
            String key = (String) params[i];
            Object value = params[i + 1];
            if (value != null) {
                sb.append(sb.length() == 0 ? "?" : "&");
                sb.append(URLEncoder.encode(key, StandardCharsets.UTF_8));
                sb.append("=");
                if (value instanceof Boolean) {
                    sb.append(((Boolean) value).toString().toLowerCase());
                } else {
                    sb.append(URLEncoder.encode(value.toString(), StandardCharsets.UTF_8));
                }
            }
        }
        return sb.toString();
    }
}

File: packages/ruby-sdk/lib/turbodocx_sdk/turbo_partner.rb
# frozen_string_literal: true

require_relative "http_client"

module TurboDocxSdk
  # TurboPartner module -- partner portal management operations.
  #
  # All methods are class-level (static pattern). Call +configure+ once,
  # then invoke any method directly on the class.
  #
  #   TurboDocxSdk::TurboPartner.configure(partner_api_key: "TDXP-...", partner_id: "...")
  #   orgs = TurboDocxSdk::TurboPartner.list_organizations
  #
  class TurboPartner
    class << self
      # Configure the TurboPartner module with partner API credentials.
      #
      # @param partner_api_key [String] Partner API key (TDXP- prefix)
      # @param partner_id [String] Partner UUID
      # @param base_url [String, nil]
      # @raise [AuthenticationError] if the partner API key is invalid
      def configure(partner_api_key:, partner_id:, base_url: nil)
        @client = HttpClient.new(
          api_key: partner_api_key,
          base_url: base_url,
          skip_sender_validation: true
        )
        @partner_id = partner_id
      end

      # ============================================
      # ORGANIZATION MANAGEMENT
      # ============================================

      # Create a new organization.
      #
      # @param request [Hash] organization data (:name, :metadata, :features)
      # @return [Hash] the created organization
      # @raise [ValidationError] on invalid request data or if not configured
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def create_organization(request)
        client = get_client
        client.post("/partner/#{partner_id}/organization", request)
      end

      # List organizations with optional pagination and search.
      #
      # @param request [Hash, nil] :limit, :offset, :search
      # @return [Hash] paginated list of organizations
      # @raise [ValidationError] if not configured and no env vars
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def list_organizations(request = nil)
        client = get_client
        client.get("/partner/#{partner_id}/organizations", to_query_params(request))
      end

      # Get detailed information about an organization.
      #
      # @param organization_id [String]
      # @return [Hash] organization details with features and tracking
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def get_organization_details(organization_id)
        client = get_client
        client.get("/partner/#{partner_id}/organizations/#{organization_id}")
      end

      # Update an organization's info.
      #
      # @param organization_id [String]
      # @param request [Hash] fields to update
      # @return [Hash] the updated organization
      # @raise [NotFoundError] if the organization does not exist
      # @raise [ValidationError] on invalid request data
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_organization_info(organization_id, request)
        client = get_client
        client.patch("/partner/#{partner_id}/organizations/#{organization_id}", request)
      end

      # Delete an organization.
      #
      # @param organization_id [String]
      # @return [Hash] deletion confirmation
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def delete_organization(organization_id)
        client = get_client
        client.delete("/partner/#{partner_id}/organizations/#{organization_id}")
      end

      # Update an organization's entitlements (features and tracking).
      #
      # @param organization_id [String]
      # @param request [Hash] :features, :tracking
      # @return [Hash] updated entitlements
      # @raise [NotFoundError] if the organization does not exist
      # @raise [ValidationError] on invalid request data
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_organization_entitlements(organization_id, request)
        client = get_client
        client.patch("/partner/#{partner_id}/organizations/#{organization_id}/entitlements", request)
      end

      # Read the TurboSign display preferences for one of the partner's organizations.
      #
      # Returns only the partner-settable preference keys, each with its effective
      # value (defaults applied for keys the org never set).
      #
      # @param organization_id [String]
      # @return [Hash] the organization's partner-settable preferences under +data.preferences+
      #   (+hideSignatureOutline+, +hideSignatureHash+, +lockedFieldsBackground+,
      #   +allowDownloadBeforeSigning+ booleans)
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def get_organization_preferences(organization_id)
        client = get_client
        client.get("/partner/#{partner_id}/organizations/#{organization_id}/preferences")
      end

      # Set TurboSign display preferences for one of the partner's organizations.
      #
      # Pass only the keys you want to change; each must be a boolean. The keys stay
      # camelCase verbatim (+hideSignatureOutline+, +hideSignatureHash+,
      # +lockedFieldsBackground+, +allowDownloadBeforeSigning+) -- they are the API
      # contract, not Ruby names.
      #
      # @param organization_id [String]
      # @param preferences [Hash] the preference keys to change (only these are sent)
      # @return [Hash] the organization's updated partner-settable preferences under +data.preferences+
      # @raise [NotFoundError] if the organization does not exist
      # @raise [ValidationError] on invalid request data
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_organization_preferences(organization_id, preferences)
        client = get_client
        # Wrap under :preferences without mutating the caller's hash.
        client.patch(
          "/partner/#{partner_id}/organizations/#{organization_id}/preferences",
          { "preferences" => preferences.dup }
        )
      end

      # ============================================
      # ORGANIZATION USER MANAGEMENT
      # ============================================

      # List users in an organization.
      #
      # @param organization_id [String]
      # @param request [Hash, nil] :limit, :offset, :search
      # @return [Hash] paginated list of users
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def list_organization_users(organization_id, request = nil)
        client = get_client
        client.get("/partner/#{partner_id}/organizations/#{organization_id}/users", to_query_params(request))
      end

      # Add a user to an organization.
      #
      # @param organization_id [String]
      # @param request [Hash] :email, :role (+"admin"+, +"contributor"+, +"user"+ or +"viewer"+ --
      #   the org role enum; the partner role +"member"+ is NOT valid here)
      # @return [Hash] the added user
      # @raise [ValidationError] on invalid request data
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def add_user_to_organization(organization_id, request)
        client = get_client
        client.post("/partner/#{partner_id}/organizations/#{organization_id}/users", request)
      end

      # Update a user's role in an organization.
      #
      # @param organization_id [String]
      # @param user_id [String]
      # @param request [Hash] :role (+"admin"+, +"contributor"+, +"user"+ or +"viewer"+)
      # @return [Hash] the updated user
      # @raise [NotFoundError] if the organization or user does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_organization_user_role(organization_id, user_id, request)
        client = get_client
        client.patch("/partner/#{partner_id}/organizations/#{organization_id}/users/#{user_id}", request)
      end

      # Remove a user from an organization.
      #
      # @param organization_id [String]
      # @param user_id [String]
      # @return [Hash] removal confirmation
      # @raise [NotFoundError] if the organization or user does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def remove_user_from_organization(organization_id, user_id)
        client = get_client
        client.delete("/partner/#{partner_id}/organizations/#{organization_id}/users/#{user_id}")
      end

      # Resend an organization invitation to a user.
      #
      # @param organization_id [String]
      # @param user_id [String]
      # @return [Hash] resend confirmation
      # @raise [NotFoundError] if the organization or user does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def resend_organization_invitation_to_user(organization_id, user_id)
        client = get_client
        client.post("/partner/#{partner_id}/organizations/#{organization_id}/users/#{user_id}/resend-invitation")
      end

      # ============================================
      # ORGANIZATION API KEY MANAGEMENT
      # ============================================

      # List API keys for an organization.
      #
      # @param organization_id [String]
      # @param request [Hash, nil] :limit, :offset, :search
      # @return [Hash] paginated list of API keys
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def list_organization_api_keys(organization_id, request = nil)
        client = get_client
        client.get("/partner/#{partner_id}/organizations/#{organization_id}/apikeys", to_query_params(request))
      end

      # Create an API key for an organization.
      #
      # @param organization_id [String]
      # @param request [Hash] :name, :role (org role enum:
      #   +"admin"+, +"contributor"+, +"user"+ or +"viewer"+)
      # @return [Hash] the created API key (includes full key value)
      # @raise [ValidationError] on invalid request data
      # @raise [NotFoundError] if the organization does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def create_organization_api_key(organization_id, request)
        client = get_client
        client.post("/partner/#{partner_id}/organizations/#{organization_id}/apikeys", request)
      end

      # Update an organization API key.
      #
      # @param organization_id [String]
      # @param api_key_id [String]
      # @param request [Hash] fields to update
      # @return [Hash] the updated API key
      # @raise [NotFoundError] if the organization or API key does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_organization_api_key(organization_id, api_key_id, request)
        client = get_client
        client.patch("/partner/#{partner_id}/organizations/#{organization_id}/apikeys/#{api_key_id}", request)
      end

      # Revoke an organization API key.
      #
      # @param organization_id [String]
      # @param api_key_id [String]
      # @return [Hash] revocation confirmation
      # @raise [NotFoundError] if the organization or API key does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def revoke_organization_api_key(organization_id, api_key_id)
        client = get_client
        client.delete("/partner/#{partner_id}/organizations/#{organization_id}/apikeys/#{api_key_id}")
      end

      # ============================================
      # PARTNER API KEY MANAGEMENT
      # ============================================

      # List partner API keys.
      #
      # @param request [Hash, nil] :limit, :offset
      # @return [Hash] paginated list of partner API keys
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def list_partner_api_keys(request = nil)
        client = get_client
        client.get("/partner/#{partner_id}/api-keys", to_query_params(request))
      end

      # Create a partner API key.
      #
      # @param request [Hash] :name, :scopes, :description
      # @return [Hash] the created partner API key (includes full key value)
      # @raise [ValidationError] on invalid request data
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def create_partner_api_key(request)
        client = get_client
        client.post("/partner/#{partner_id}/api-keys", request)
      end

      # Update a partner API key.
      #
      # @param key_id [String]
      # @param request [Hash] fields to update
      # @return [Hash] the updated partner API key
      # @raise [NotFoundError] if the API key does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_partner_api_key(key_id, request)
        client = get_client
        client.patch("/partner/#{partner_id}/api-keys/#{key_id}", request)
      end

      # Revoke a partner API key.
      #
      # @param key_id [String]
      # @return [Hash] revocation confirmation
      # @raise [NotFoundError] if the API key does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def revoke_partner_api_key(key_id)
        client = get_client
        client.delete("/partner/#{partner_id}/api-keys/#{key_id}")
      end

      # ============================================
      # PARTNER USER MANAGEMENT
      # ============================================

      # List partner portal users.
      #
      # @param request [Hash, nil] :limit, :offset, :search
      # @return [Hash] paginated list of partner users
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def list_partner_portal_users(request = nil)
        client = get_client
        client.get("/partner/#{partner_id}/users", to_query_params(request))
      end

      # Add a user to the partner portal.
      #
      # @param request [Hash] :email, :role (partner role enum: +"admin"+, +"member"+ or
      #   +"viewer"+ -- the org roles +"contributor"+/+"user"+ are NOT valid here),
      #   :permissions. When :permissions is supplied it must carry ALL SEVEN keys --
      #   +canManageOrgs+, +canManageOrgUsers+, +canManagePartnerUsers+,
      #   +canManageOrgAPIKeys+, +canManagePartnerAPIKeys+, +canUpdateEntitlements+,
      #   +canViewAuditLogs+ (booleans). A partial permissions object is a 400.
      # @return [Hash] the added user
      # @raise [ValidationError] on invalid request data
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def add_user_to_partner_portal(request)
        client = get_client
        client.post("/partner/#{partner_id}/users", request)
      end

      # Update a partner user's permissions.
      #
      # The :permissions object itself is optional, but there is no partial update:
      # if you send it, send all seven keys or the backend returns a 400.
      #
      # @param user_id [String]
      # @param request [Hash] :role (+"admin"+, +"member"+ or +"viewer"+), :permissions
      #   (ALL SEVEN of +canManageOrgs+, +canManageOrgUsers+, +canManagePartnerUsers+,
      #   +canManageOrgAPIKeys+, +canManagePartnerAPIKeys+, +canUpdateEntitlements+,
      #   +canViewAuditLogs+)
      # @return [Hash] the updated user
      # @raise [NotFoundError] if the user does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def update_partner_user_permissions(user_id, request)
        client = get_client
        client.patch("/partner/#{partner_id}/users/#{user_id}", request)
      end

      # Remove a user from the partner portal.
      #
      # @param user_id [String]
      # @return [Hash] removal confirmation
      # @raise [NotFoundError] if the user does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def remove_user_from_partner_portal(user_id)
        client = get_client
        client.delete("/partner/#{partner_id}/users/#{user_id}")
      end

      # Resend a partner portal invitation to a user.
      #
      # @param user_id [String]
      # @return [Hash] resend confirmation
      # @raise [NotFoundError] if the user does not exist
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def resend_partner_portal_invitation_to_user(user_id)
        client = get_client
        client.post("/partner/#{partner_id}/users/#{user_id}/resend-invitation")
      end

      # ============================================
      # AUDIT LOGS
      # ============================================

      # Get partner audit logs with optional filters.
      #
      # @param request [Hash, nil] :limit, :offset, :action, :resourceType, :success, :startDate, :endDate
      # @return [Hash] paginated list of audit log entries
      # @raise [AuthenticationError] on invalid credentials
      # @raise [NetworkError] on connection failure
      def get_partner_audit_logs(request = nil)
        client = get_client
        client.get("/partner/#{partner_id}/audit-logs", to_query_params(request))
      end

      private

      def get_client
        @client ||= begin
          partner_api_key = ENV["TURBODOCX_PARTNER_API_KEY"]
          pid = ENV["TURBODOCX_PARTNER_ID"]
          unless partner_api_key && pid
            raise ValidationError, "TurboPartner must be configured before use. Call TurboPartner.configure() " \
                  "or set TURBODOCX_PARTNER_API_KEY and TURBODOCX_PARTNER_ID environment variables."
          end
          configure(partner_api_key: partner_api_key, partner_id: pid)
          @client
        end
      end

      def partner_id
        @partner_id || begin
          get_client # triggers auto-init
          @partner_id
        end
      end

      def to_query_params(request)
        return nil if request.nil? || request.empty?

        params = {}
        request.each do |key, value|
          str_key = key.to_s
          next if value.nil?

          case value
          when true, false
            params[str_key] = value ? "true" : "false"
          when Array
            # Preserve as a string array so HttpClient#build_url emits repeated
            # keys (?k=a&k=b) rather than a JSON-stringified blob. Mirrors
            # TurboQuote#to_query_params.
            params[str_key] = value.map(&:to_s)
          else
            params[str_key] = value.to_s
          end
        end
        params.empty? ? nil : params
      end
    end
  end
end
